<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_37) on Sun Dec 02 01:32:55 EST 2012 -->
<TITLE>
Client
</TITLE>

<META NAME="date" CONTENT="2012-12-02">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Client";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/CacheGCHook.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../../index.html?edu/psu/compbio/seqcode/gse/projects/readdb/Client.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Client.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
edu.psu.compbio.seqcode.gse.projects.readdb</FONT>
<BR>
Class Client</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../../../../resources/inherit.gif" ALT="extended by "><B>edu.psu.compbio.seqcode.gse.projects.readdb.Client</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html" title="interface in edu.psu.compbio.seqcode.gse.projects.readdb">ReadOnlyClient</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>Client</B><DT>extends java.lang.Object<DT>implements <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html" title="interface in edu.psu.compbio.seqcode.gse.projects.readdb">ReadOnlyClient</A></DL>
</PRE>

<P>
<p>API for remote access to the readdb server.

  <p>Calls throw IOException on network errors.
  Calls throw ClientException on other errors (authentication, authorization, invalid request ,etc.

  <p>Client generally assumes that the hit positions are the 5' end of the
  hit. 

 <p>Client IS NOT REENTRANT.  Do not overlap calls to a single Client object.

 <p>Most method parameters that are object types (eg Integer, Boolean) are optional.  If a null value
 is passed then no filtering is done based on that parameter.  

 <p>Standard parameters shared across methods:
 <ul>
 <li> alignid is the name of the alignment.
 <li> isPaired specifies whether to work on single-ended reads (false) or paired-end reads (true)
 <li> isLeft is isPaired is true, then isLeft specifies whether to work on the left read (true)(
     or right read (false) of the pair
 <li> plusStrand specifies whether to return only reads on the plus strand (true) or minus strand (false). null
     means that reads on both strands should be returned.
 <li> minWeight specifies the minimum weight of reads (or read pairs) to be returned or included in 
     the histogram
 <li> start, stop specify the lowest (inclusive) and highest (exclusive) coordinates of reads
     that should be included in the results
 </ul>
<P>

<P>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#SaslMechanisms">SaslMechanisms</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#Client()">Client</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates the default connection
 as specified by ~/.readdb_passwd or a readdb_passwd found in the classpath
 Must have keys hostname, port, username, and passwd in a format
 that java.util.PropertyResourceBundle can read</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#Client(java.lang.String, int, java.lang.String, java.lang.String)">Client</A></B>(java.lang.String&nbsp;hostname,
       int&nbsp;portnum,
       java.lang.String&nbsp;username,
       java.lang.String&nbsp;passwd)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Connects to a Readdb server on the specified host and port using the specified 
 username and password.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#addToGroup(java.lang.String, java.lang.String)">addToGroup</A></B>(java.lang.String&nbsp;princ,
           java.lang.String&nbsp;group)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds the specified user (princ) to a group.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#checksort(java.lang.String, int)">checksort</A></B>(java.lang.String&nbsp;align,
          int&nbsp;chrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This was to fix a bug in the server.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#close()">close</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Closes this connection to the server.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#deleteAlignment(java.lang.String, java.lang.Boolean)">deleteAlignment</A></B>(java.lang.String&nbsp;alignid,
                java.lang.Boolean&nbsp;isPaired)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Deletes an alignment (all chromosomes).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#exists(java.lang.String)">exists</A></B>(java.lang.String&nbsp;alignid)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns true if the alignment and chromosome exist and are accessible
 to the user.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getACL(java.lang.String)">getACL</A></B>(java.lang.String&nbsp;alignid)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a Map from READ, WRITE, and ADMIN to lists of principals that have those privileges on the specified alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set&lt;java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getChroms(java.lang.String, boolean, java.lang.Boolean)">getChroms</A></B>(java.lang.String&nbsp;alignid,
          boolean&nbsp;isPaired,
          java.lang.Boolean&nbsp;isLeft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the set of chromosomes that exist for this alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getCount(java.lang.String, boolean, java.lang.Boolean, java.lang.Boolean)">getCount</A></B>(java.lang.String&nbsp;alignid,
         boolean&nbsp;isPaired,
         java.lang.Boolean&nbsp;isLeft,
         java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the total number of hits in this alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getCount(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)">getCount</A></B>(java.lang.String&nbsp;alignid,
         int&nbsp;chromid,
         boolean&nbsp;paired,
         java.lang.Integer&nbsp;start,
         java.lang.Integer&nbsp;stop,
         java.lang.Float&nbsp;minWeight,
         java.lang.Boolean&nbsp;isLeft,
         java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns the total number of hits on the specified chromosome in the alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getHistogram(java.util.Collection, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getHistogram</A></B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
             int&nbsp;chromid,
             boolean&nbsp;paired,
             boolean&nbsp;doReadExtension,
             int&nbsp;binsize,
             java.lang.Integer&nbsp;start,
             java.lang.Integer&nbsp;stop,
             java.lang.Float&nbsp;minWeight,
             java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getHistogram(java.util.Collection, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getHistogram</A></B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
             int&nbsp;chromid,
             boolean&nbsp;paired,
             boolean&nbsp;doReadExtension,
             int&nbsp;binsize,
             int&nbsp;dedup,
             java.lang.Integer&nbsp;start,
             java.lang.Integer&nbsp;stop,
             java.lang.Float&nbsp;minWeight,
             java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getHistogram(java.lang.String, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getHistogram</A></B>(java.lang.String&nbsp;alignid,
             int&nbsp;chromid,
             boolean&nbsp;paired,
             boolean&nbsp;doReadExtension,
             int&nbsp;binsize,
             java.lang.Integer&nbsp;start,
             java.lang.Integer&nbsp;stop,
             java.lang.Float&nbsp;minWeight,
             java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns a TreeMap from positions to counts representing a histogram
 of the hits in a range with the specified binsize.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getHistogram(java.lang.String, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, boolean)">getHistogram</A></B>(java.lang.String&nbsp;alignid,
             int&nbsp;chromid,
             boolean&nbsp;paired,
             boolean&nbsp;doReadExtension,
             int&nbsp;binsize,
             int&nbsp;dedup,
             java.lang.Integer&nbsp;start,
             java.lang.Integer&nbsp;stop,
             java.lang.Float&nbsp;minWeight,
             java.lang.Boolean&nbsp;plusStrand,
             boolean&nbsp;isLeft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/PairedHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">PairedHit</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getPairedHits(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getPairedHits</A></B>(java.lang.String&nbsp;alignid,
              int&nbsp;chromid,
              boolean&nbsp;isLeft,
              java.lang.Integer&nbsp;start,
              java.lang.Integer&nbsp;stop,
              java.lang.Float&nbsp;minWeight,
              java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getPositions(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)">getPositions</A></B>(java.lang.String&nbsp;alignid,
             int&nbsp;chromid,
             boolean&nbsp;paired,
             java.lang.Integer&nbsp;start,
             java.lang.Integer&nbsp;stop,
             java.lang.Float&nbsp;minWeight,
             java.lang.Boolean&nbsp;isLeft,
             java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns the sorted (ascending order) hit positions in the specified range of a chromosome,alignment pair.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/SingleHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">SingleHit</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getSingleHits(java.lang.String, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getSingleHits</A></B>(java.lang.String&nbsp;alignid,
              int&nbsp;chromid,
              java.lang.Integer&nbsp;start,
              java.lang.Integer&nbsp;stop,
              java.lang.Float&nbsp;minWeight,
              java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getWeight(java.lang.String, boolean, java.lang.Boolean, java.lang.Boolean)">getWeight</A></B>(java.lang.String&nbsp;alignid,
          boolean&nbsp;isPaired,
          java.lang.Boolean&nbsp;isLeft,
          java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the sum of the weights of all hits in this alignment</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getWeight(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)">getWeight</A></B>(java.lang.String&nbsp;alignid,
          int&nbsp;chromid,
          boolean&nbsp;paired,
          java.lang.Integer&nbsp;start,
          java.lang.Integer&nbsp;stop,
          java.lang.Float&nbsp;minWeight,
          java.lang.Boolean&nbsp;isLeft,
          java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns the total weight on the specified chromosome in this alignment</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getWeightHistogram(java.util.Collection, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getWeightHistogram</A></B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
                   int&nbsp;chromid,
                   boolean&nbsp;paired,
                   boolean&nbsp;doReadExtension,
                   int&nbsp;binsize,
                   java.lang.Integer&nbsp;start,
                   java.lang.Integer&nbsp;stop,
                   java.lang.Float&nbsp;minWeight,
                   java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getWeightHistogram(java.util.Collection, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getWeightHistogram</A></B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
                   int&nbsp;chromid,
                   boolean&nbsp;paired,
                   boolean&nbsp;doReadExtension,
                   int&nbsp;binsize,
                   int&nbsp;dedup,
                   java.lang.Integer&nbsp;start,
                   java.lang.Integer&nbsp;stop,
                   java.lang.Float&nbsp;minWeight,
                   java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getWeightHistogram(java.lang.String, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getWeightHistogram</A></B>(java.lang.String&nbsp;alignid,
                   int&nbsp;chromid,
                   boolean&nbsp;paired,
                   boolean&nbsp;doReadExtension,
                   int&nbsp;binsize,
                   java.lang.Integer&nbsp;start,
                   java.lang.Integer&nbsp;stop,
                   java.lang.Float&nbsp;minWeight,
                   java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getWeightHistogram(java.lang.String, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, boolean)">getWeightHistogram</A></B>(java.lang.String&nbsp;alignid,
                   int&nbsp;chromid,
                   boolean&nbsp;paired,
                   boolean&nbsp;doReadExtension,
                   int&nbsp;binsize,
                   int&nbsp;dedup,
                   java.lang.Integer&nbsp;start,
                   java.lang.Integer&nbsp;stop,
                   java.lang.Float&nbsp;minWeight,
                   java.lang.Boolean&nbsp;plusStrand,
                   boolean&nbsp;isLeft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;float[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#getWeightsRange(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)">getWeightsRange</A></B>(java.lang.String&nbsp;alignid,
                int&nbsp;chromid,
                boolean&nbsp;paired,
                java.lang.Integer&nbsp;start,
                java.lang.Integer&nbsp;stop,
                java.lang.Float&nbsp;minWeight,
                java.lang.Boolean&nbsp;isLeft,
                java.lang.Boolean&nbsp;plusStrand)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns the hit weights in the specified range of a chromosome,alignment pair.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#printErrors(boolean)">printErrors</A></B>(boolean&nbsp;b)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Determines whether the client will print error messages to STDERR.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#reIndex(java.lang.String, int, boolean)">reIndex</A></B>(java.lang.String&nbsp;align,
        int&nbsp;chrom,
        boolean&nbsp;paired)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this was to fix a bug in the server.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#setACL(java.lang.String, java.util.Set)">setACL</A></B>(java.lang.String&nbsp;alignid,
       java.util.Set&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ACLChangeEntry.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ACLChangeEntry</A>&gt;&nbsp;changes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Applies the specified ACLChangeEntry objects to the acl for this experiment/chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#shutdown()">shutdown</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tells the server to shut itself down.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#storePaired(java.lang.String, java.util.List)">storePaired</A></B>(java.lang.String&nbsp;alignid,
            java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/PairedHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">PairedHit</A>&gt;&nbsp;allhits)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Stores a set of PairedHit objects (representing an paired-ended read
 aligned to a genome) in the specified alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/Client.html#storeSingle(java.lang.String, java.util.List)">storeSingle</A></B>(java.lang.String&nbsp;alignid,
            java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/SingleHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">SingleHit</A>&gt;&nbsp;allhits)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Stores a set of SingleHit objects (representing an un-paired or single-ended read
 aligned to a genome) in the specified alignment.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="SaslMechanisms"><!-- --></A><H3>
SaslMechanisms</H3>
<PRE>
public static final java.lang.String[] <B>SaslMechanisms</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Client(java.lang.String, int, java.lang.String, java.lang.String)"><!-- --></A><H3>
Client</H3>
<PRE>
public <B>Client</B>(java.lang.String&nbsp;hostname,
              int&nbsp;portnum,
              java.lang.String&nbsp;username,
              java.lang.String&nbsp;passwd)
       throws java.io.IOException,
              <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Connects to a Readdb server on the specified host and port using the specified 
 username and password.
<P>
<DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - on network errors
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE> - if the client cannot authenticate to the server</DL>
</DL>
<HR>

<A NAME="Client()"><!-- --></A><H3>
Client</H3>
<PRE>
public <B>Client</B>()
       throws java.io.IOException,
              <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Creates the default connection
 as specified by ~/.readdb_passwd or a readdb_passwd found in the classpath
 Must have keys hostname, port, username, and passwd in a format
 that java.util.PropertyResourceBundle can read
<P>
<DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - on network errors
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE> - if the client cannot authenticate to the server</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="printErrors(boolean)"><!-- --></A><H3>
printErrors</H3>
<PRE>
public void <B>printErrors</B>(boolean&nbsp;b)</PRE>
<DL>
<DD>Determines whether the client will print error messages to STDERR.  Useful for debugging 
 but may produce unwanted screen output.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="shutdown()"><!-- --></A><H3>
shutdown</H3>
<PRE>
public void <B>shutdown</B>()
              throws java.io.IOException,
                     <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Tells the server to shut itself down.  Use this to stop the server process.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - on network errors
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE> - if the user isn't authorized to shut the server down.</DL>
</DD>
</DL>
<HR>

<A NAME="reIndex(java.lang.String, int, boolean)"><!-- --></A><H3>
reIndex</H3>
<PRE>
public void <B>reIndex</B>(java.lang.String&nbsp;align,
                    int&nbsp;chrom,
                    boolean&nbsp;paired)
             throws java.io.IOException,
                    <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>this was to fix a bug in the server.  You shouldn't need it for general use.
 Regenerate the index for this alignment and chromosome
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="checksort(java.lang.String, int)"><!-- --></A><H3>
checksort</H3>
<PRE>
public void <B>checksort</B>(java.lang.String&nbsp;align,
                      int&nbsp;chrom)
               throws java.io.IOException,
                      <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>This was to fix a bug in the server.  You shouldn't need it for general use.
 Resort the hits for a single-ended alignment and regenerate the index.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="storeSingle(java.lang.String, java.util.List)"><!-- --></A><H3>
storeSingle</H3>
<PRE>
public void <B>storeSingle</B>(java.lang.String&nbsp;alignid,
                        java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/SingleHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">SingleHit</A>&gt;&nbsp;allhits)
                 throws java.io.IOException,
                        <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Stores a set of SingleHit objects (representing an un-paired or single-ended read
 aligned to a genome) in the specified alignment.  The hits are appended
 to any hits that have already been stored in the alignment.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="storePaired(java.lang.String, java.util.List)"><!-- --></A><H3>
storePaired</H3>
<PRE>
public void <B>storePaired</B>(java.lang.String&nbsp;alignid,
                        java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/PairedHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">PairedHit</A>&gt;&nbsp;allhits)
                 throws java.io.IOException,
                        <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Stores a set of PairedHit objects (representing an paired-ended read
 aligned to a genome) in the specified alignment.  The hits are appended
 to any hits that have already been stored in the alignment
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="exists(java.lang.String)"><!-- --></A><H3>
exists</H3>
<PRE>
public boolean <B>exists</B>(java.lang.String&nbsp;alignid)
               throws java.io.IOException</PRE>
<DL>
<DD>Returns true if the alignment and chromosome exist and are accessible
 to the user.  Returns false if they don't exist or if they aren't accessible
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="deleteAlignment(java.lang.String, java.lang.Boolean)"><!-- --></A><H3>
deleteAlignment</H3>
<PRE>
public void <B>deleteAlignment</B>(java.lang.String&nbsp;alignid,
                            java.lang.Boolean&nbsp;isPaired)
                     throws java.io.IOException,
                            <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Deletes an alignment (all chromosomes).  isPaired specifies whether to delete
 the paired or single ended reads.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getChroms(java.lang.String, boolean, java.lang.Boolean)"><!-- --></A><H3>
getChroms</H3>
<PRE>
public java.util.Set&lt;java.lang.Integer&gt; <B>getChroms</B>(java.lang.String&nbsp;alignid,
                                                  boolean&nbsp;isPaired,
                                                  java.lang.Boolean&nbsp;isLeft)
                                           throws java.io.IOException,
                                                  <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Returns the set of chromosomes that exist for this alignment.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getCount(java.lang.String, boolean, java.lang.Boolean, java.lang.Boolean)"><!-- --></A><H3>
getCount</H3>
<PRE>
public int <B>getCount</B>(java.lang.String&nbsp;alignid,
                    boolean&nbsp;isPaired,
                    java.lang.Boolean&nbsp;isLeft,
                    java.lang.Boolean&nbsp;plusStrand)
             throws java.io.IOException,
                    <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Returns the total number of hits in this alignment.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getWeight(java.lang.String, boolean, java.lang.Boolean, java.lang.Boolean)"><!-- --></A><H3>
getWeight</H3>
<PRE>
public double <B>getWeight</B>(java.lang.String&nbsp;alignid,
                        boolean&nbsp;isPaired,
                        java.lang.Boolean&nbsp;isLeft,
                        java.lang.Boolean&nbsp;plusStrand)
                 throws java.io.IOException,
                        <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Returns the sum of the weights of all hits in this alignment
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getCount(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)"><!-- --></A><H3>
getCount</H3>
<PRE>
public int <B>getCount</B>(java.lang.String&nbsp;alignid,
                    int&nbsp;chromid,
                    boolean&nbsp;paired,
                    java.lang.Integer&nbsp;start,
                    java.lang.Integer&nbsp;stop,
                    java.lang.Float&nbsp;minWeight,
                    java.lang.Boolean&nbsp;isLeft,
                    java.lang.Boolean&nbsp;plusStrand)
             throws java.io.IOException,
                    <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>returns the total number of hits on the specified chromosome in the alignment.
 Any of the object parameters can be set to null to specify "no value"
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getWeight(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)"><!-- --></A><H3>
getWeight</H3>
<PRE>
public double <B>getWeight</B>(java.lang.String&nbsp;alignid,
                        int&nbsp;chromid,
                        boolean&nbsp;paired,
                        java.lang.Integer&nbsp;start,
                        java.lang.Integer&nbsp;stop,
                        java.lang.Float&nbsp;minWeight,
                        java.lang.Boolean&nbsp;isLeft,
                        java.lang.Boolean&nbsp;plusStrand)
                 throws java.io.IOException,
                        <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>returns the total weight on the specified chromosome in this alignment
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getPositions(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)"><!-- --></A><H3>
getPositions</H3>
<PRE>
public int[] <B>getPositions</B>(java.lang.String&nbsp;alignid,
                          int&nbsp;chromid,
                          boolean&nbsp;paired,
                          java.lang.Integer&nbsp;start,
                          java.lang.Integer&nbsp;stop,
                          java.lang.Float&nbsp;minWeight,
                          java.lang.Boolean&nbsp;isLeft,
                          java.lang.Boolean&nbsp;plusStrand)
                   throws java.io.IOException,
                          <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>returns the sorted (ascending order) hit positions in the specified range of a chromosome,alignment pair.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getWeightsRange(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, java.lang.Boolean)"><!-- --></A><H3>
getWeightsRange</H3>
<PRE>
public float[] <B>getWeightsRange</B>(java.lang.String&nbsp;alignid,
                               int&nbsp;chromid,
                               boolean&nbsp;paired,
                               java.lang.Integer&nbsp;start,
                               java.lang.Integer&nbsp;stop,
                               java.lang.Float&nbsp;minWeight,
                               java.lang.Boolean&nbsp;isLeft,
                               java.lang.Boolean&nbsp;plusStrand)
                        throws java.io.IOException,
                               <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>returns the hit weights in the specified range of a chromosome,alignment pair.  The weights
 will be in the same order as the sorted positions returned by getPositions()
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getSingleHits(java.lang.String, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getSingleHits</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/SingleHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">SingleHit</A>&gt; <B>getSingleHits</B>(java.lang.String&nbsp;alignid,
                                               int&nbsp;chromid,
                                               java.lang.Integer&nbsp;start,
                                               java.lang.Integer&nbsp;stop,
                                               java.lang.Float&nbsp;minWeight,
                                               java.lang.Boolean&nbsp;plusStrand)
                                        throws java.io.IOException,
                                               <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getPairedHits(java.lang.String, int, boolean, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getPairedHits</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/PairedHit.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">PairedHit</A>&gt; <B>getPairedHits</B>(java.lang.String&nbsp;alignid,
                                               int&nbsp;chromid,
                                               boolean&nbsp;isLeft,
                                               java.lang.Integer&nbsp;start,
                                               java.lang.Integer&nbsp;stop,
                                               java.lang.Float&nbsp;minWeight,
                                               java.lang.Boolean&nbsp;plusStrand)
                                        throws java.io.IOException,
                                               <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getHistogram(java.lang.String, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt; <B>getHistogram</B>(java.lang.String&nbsp;alignid,
                                                                           int&nbsp;chromid,
                                                                           boolean&nbsp;paired,
                                                                           boolean&nbsp;doReadExtension,
                                                                           int&nbsp;binsize,
                                                                           java.lang.Integer&nbsp;start,
                                                                           java.lang.Integer&nbsp;stop,
                                                                           java.lang.Float&nbsp;minWeight,
                                                                           java.lang.Boolean&nbsp;plusStrand)
                                                                    throws java.io.IOException,
                                                                           <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>returns a TreeMap from positions to counts representing a histogram
 of the hits in a range with the specified binsize.  Bins with a count
 of zero are not included in the output.
 
 Ex: getHistgram("foo","1+",1,100,10)
     6: 5
    16: 0
    36: 30

 minweight is the minimum weight for reads to be included in the histogram.

 dedup is the limit on how many times reads with any given 5' position will be counted.
  A value of zero means no limit.  A limit of, eg, 2, means that at most two reads at
  any 5' position will be included in the output.  For weighted histograms, the choice of reads
  included is unspecified.  For methods that operate on a set of alignments, this many
  reads from each alignment will be included.

 Normally, a read is only counted in a single bin as defined by its position (generally
 the 5' end of the read).  A non-zero read-Extension counts the read in any
 bin that you cross within that many bases of the read's position.  A negative value
 goes backwards (smaller coordinates) and a larger value goes forward.  You need
 to get the sign right depending on the strandedness of the chromosome that you're
 working with.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html#getHistogram(java.lang.String, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getHistogram</A></CODE> in interface <CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html" title="interface in edu.psu.compbio.seqcode.gse.projects.readdb">ReadOnlyClient</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getHistogram(java.lang.String, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, boolean)"><!-- --></A><H3>
getHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt; <B>getHistogram</B>(java.lang.String&nbsp;alignid,
                                                                           int&nbsp;chromid,
                                                                           boolean&nbsp;paired,
                                                                           boolean&nbsp;doReadExtension,
                                                                           int&nbsp;binsize,
                                                                           int&nbsp;dedup,
                                                                           java.lang.Integer&nbsp;start,
                                                                           java.lang.Integer&nbsp;stop,
                                                                           java.lang.Float&nbsp;minWeight,
                                                                           java.lang.Boolean&nbsp;plusStrand,
                                                                           boolean&nbsp;isLeft)
                                                                    throws java.io.IOException,
                                                                           <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getWeightHistogram(java.lang.String, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getWeightHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt; <B>getWeightHistogram</B>(java.lang.String&nbsp;alignid,
                                                                               int&nbsp;chromid,
                                                                               boolean&nbsp;paired,
                                                                               boolean&nbsp;doReadExtension,
                                                                               int&nbsp;binsize,
                                                                               java.lang.Integer&nbsp;start,
                                                                               java.lang.Integer&nbsp;stop,
                                                                               java.lang.Float&nbsp;minWeight,
                                                                               java.lang.Boolean&nbsp;plusStrand)
                                                                        throws java.io.IOException,
                                                                               <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html#getWeightHistogram(java.lang.String, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)">getWeightHistogram</A></CODE> in interface <CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html" title="interface in edu.psu.compbio.seqcode.gse.projects.readdb">ReadOnlyClient</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getWeightHistogram(java.lang.String, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean, boolean)"><!-- --></A><H3>
getWeightHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt; <B>getWeightHistogram</B>(java.lang.String&nbsp;alignid,
                                                                               int&nbsp;chromid,
                                                                               boolean&nbsp;paired,
                                                                               boolean&nbsp;doReadExtension,
                                                                               int&nbsp;binsize,
                                                                               int&nbsp;dedup,
                                                                               java.lang.Integer&nbsp;start,
                                                                               java.lang.Integer&nbsp;stop,
                                                                               java.lang.Float&nbsp;minWeight,
                                                                               java.lang.Boolean&nbsp;plusStrand,
                                                                               boolean&nbsp;isLeft)
                                                                        throws java.io.IOException,
                                                                               <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getHistogram(java.util.Collection, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt; <B>getHistogram</B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
                                                                           int&nbsp;chromid,
                                                                           boolean&nbsp;paired,
                                                                           boolean&nbsp;doReadExtension,
                                                                           int&nbsp;binsize,
                                                                           java.lang.Integer&nbsp;start,
                                                                           java.lang.Integer&nbsp;stop,
                                                                           java.lang.Float&nbsp;minWeight,
                                                                           java.lang.Boolean&nbsp;plusStrand)
                                                                    throws java.io.IOException,
                                                                           <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getHistogram(java.util.Collection, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Integer&gt; <B>getHistogram</B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
                                                                           int&nbsp;chromid,
                                                                           boolean&nbsp;paired,
                                                                           boolean&nbsp;doReadExtension,
                                                                           int&nbsp;binsize,
                                                                           int&nbsp;dedup,
                                                                           java.lang.Integer&nbsp;start,
                                                                           java.lang.Integer&nbsp;stop,
                                                                           java.lang.Float&nbsp;minWeight,
                                                                           java.lang.Boolean&nbsp;plusStrand)
                                                                    throws java.io.IOException,
                                                                           <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getWeightHistogram(java.util.Collection, int, boolean, boolean, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getWeightHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt; <B>getWeightHistogram</B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
                                                                               int&nbsp;chromid,
                                                                               boolean&nbsp;paired,
                                                                               boolean&nbsp;doReadExtension,
                                                                               int&nbsp;binsize,
                                                                               java.lang.Integer&nbsp;start,
                                                                               java.lang.Integer&nbsp;stop,
                                                                               java.lang.Float&nbsp;minWeight,
                                                                               java.lang.Boolean&nbsp;plusStrand)
                                                                        throws java.io.IOException,
                                                                               <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getWeightHistogram(java.util.Collection, int, boolean, boolean, int, int, java.lang.Integer, java.lang.Integer, java.lang.Float, java.lang.Boolean)"><!-- --></A><H3>
getWeightHistogram</H3>
<PRE>
public java.util.TreeMap&lt;java.lang.Integer,java.lang.Float&gt; <B>getWeightHistogram</B>(java.util.Collection&lt;java.lang.String&gt;&nbsp;alignids,
                                                                               int&nbsp;chromid,
                                                                               boolean&nbsp;paired,
                                                                               boolean&nbsp;doReadExtension,
                                                                               int&nbsp;binsize,
                                                                               int&nbsp;dedup,
                                                                               java.lang.Integer&nbsp;start,
                                                                               java.lang.Integer&nbsp;stop,
                                                                               java.lang.Float&nbsp;minWeight,
                                                                               java.lang.Boolean&nbsp;plusStrand)
                                                                        throws java.io.IOException,
                                                                               <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getACL(java.lang.String)"><!-- --></A><H3>
getACL</H3>
<PRE>
public java.util.Map&lt;java.lang.String,java.util.Set&lt;java.lang.String&gt;&gt; <B>getACL</B>(java.lang.String&nbsp;alignid)
                                                                       throws java.io.IOException,
                                                                              <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Returns a Map from READ, WRITE, and ADMIN to lists of principals that have those privileges on the specified alignment.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="setACL(java.lang.String, java.util.Set)"><!-- --></A><H3>
setACL</H3>
<PRE>
public void <B>setACL</B>(java.lang.String&nbsp;alignid,
                   java.util.Set&lt;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ACLChangeEntry.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ACLChangeEntry</A>&gt;&nbsp;changes)
            throws java.io.IOException,
                   <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Applies the specified ACLChangeEntry objects to the acl for this experiment/chromosome.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addToGroup(java.lang.String, java.lang.String)"><!-- --></A><H3>
addToGroup</H3>
<PRE>
public void <B>addToGroup</B>(java.lang.String&nbsp;princ,
                       java.lang.String&nbsp;group)
                throws java.io.IOException,
                       <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></PRE>
<DL>
<DD>Adds the specified user (princ) to a group.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb">ClientException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="close()"><!-- --></A><H3>
close</H3>
<PRE>
public void <B>close</B>()</PRE>
<DL>
<DD>Closes this connection to the server.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html#close()">close</A></CODE> in interface <CODE><A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ReadOnlyClient.html" title="interface in edu.psu.compbio.seqcode.gse.projects.readdb">ReadOnlyClient</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/CacheGCHook.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../../edu/psu/compbio/seqcode/gse/projects/readdb/ClientException.html" title="class in edu.psu.compbio.seqcode.gse.projects.readdb"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../../index.html?edu/psu/compbio/seqcode/gse/projects/readdb/Client.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Client.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
